src/ui/components.tsx

```diff
@@ -243,107 +244,48 @@ const SingleTag = (props: {
 }) => {
   const { onChange, autocompleter } = props;
   const value: TagValue = props.value === undefined ? { operator: 'is' } : props.value;
+  const groupId = value.group;
 
-  const [tagOptions, setTagOptions] = React.useState<Array<SelectableValue<string>>>([]);
-  const [groupOptions, setGroupOptions] = React.useState<Array<SelectableValue<string>>>([]);
-  // we just use this state for notifying useEffect of the changed data. TODO: perhaps there is a better way?
-  const [groupId, setGroupId] = React.useState<string | undefined>(value.group);
-
-  React.useEffect(() => {
-    async function inner() {
-      // TODO: use teardown
-      setGroupOptions(await autocompleter('', 'groups', {}));
-    }
-    inner();
-  }, [autocompleter]);
-
-  React.useEffect(() => {
-    async function inner() {
-      // TODO: use teardown
-      if (groupId === undefined) {
-        return;
-      }
-      setTagOptions(await autocompleter('', 'choices', { groupId: groupId }));
-    }
-    inner();
-  }, [autocompleter, groupId]);
-
-  //  TODO: use teardown above!
-  //  React.useEffect(() => {
-  //    let useAsyncResult = true; // https://beta.reactjs.org/apis/react/useEffect#fetching-data-with-effects
-  //    async function inner() {
-  //      const options = await autocompleter('');
-  //      if (useAsyncResult) {
-  //        setOptions(options);
-  //      }
-  //    }
-  //    return () => {
-  //      useAsyncResult = false;
-  //    };
-  //  }, [autocompleter, label]);
-
-  const tagOperators = [
-    { value: 'is', label: '=' },
-    { value: 'is not', label: '≠' },
-  ];
-
-  const findOperatorSelectable = function (value: string | undefined): SelectableValue<string> | undefined {
-    for (const operator of tagOperators) {
-      if (operator.value === value || (operator.value === 'is' && value === undefined)) {
-        return operator;
-      }
-    }
-    return undefined;
-  };
-
-  const findGroupSelectable = function (value: string | undefined): SelectableValue<string> | null {
-    if (groupOptions === undefined) {
-      return null;
-    }
-    for (const group of groupOptions) {
-      if (group.value === value) {
-        return group;
-      }
-    }
-    return null;
-  };
-
-  const findTagSelectable = function (value: string | undefined): SelectableValue<string> | null {
-    // TODO: c&p from above!
-    if (tagOptions === undefined) {
-      return null;
-    }
-    for (const tag of tagOptions) {
-      if (tag.value === value) {
-        return tag;
-      }
-    }
-    return null;
-  };
+  const groupAutocompleter = React.useCallback(
+    (prefix: string) => autocompleter(prefix, 'groups', {}),
+    [autocompleter]
+  );
+  const operatorAutocompleter = React.useCallback(
+    (prefix: string) =>
+      Promise.resolve([
+        { value: 'is', label: '=' },
+        { value: 'is not', label: '≠' },
+      ]),
+    []
+  );
```

Comment:
I don't think the useCallback is needed here. No need to cache something that is always constant.